lib/views/post/widgets/vote_button.dart

```diff
@@ -0,0 +1,170 @@
+  Color lastColor;
+  IconData icon;
+
+  onStateChange(VoteAnimation voteAnimation) {
+    setState(() {
+      lastColor = color;
+      color = _getColorByAnimation(voteAnimation);
+
+      if (voteAnimation == VoteAnimation.voteFocused ||
+          voteAnimation == VoteAnimation.voteUnfocused) {
+        if (widget.type == VoteButtonType.favorite) {
+          if (widget.animationService.initialVote == Vote.favorite)
+            favoriteController.reverse();
+          else
+            favoriteController.forward();
+        } else
+          rotationController.forward();
+      } else if (voteAnimation == VoteAnimation.clearFocused ||
+          voteAnimation == VoteAnimation.clearUnfocused) {
+        if (widget.type == VoteButtonType.favorite) {
+          if (widget.animationService.initialVote != Vote.favorite)
+            favoriteController.reverse();
+          else
+            favoriteController.forward();
+        } else
+          rotationController.reverse();
+      }
+
+      if (lastColor != color) {
+        colorTween =
+            ColorTween(begin: lastColor, end: color).animate(colorController);
+        colorController.forward(from: 0);
+      }
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    lastColor = color = _getInitialColor();
+    rotationController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    colorController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    favoriteController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    widget.animationService.addButtonStateListener(widget.type, onStateChange);
+    icon = widget.type == VoteButtonType.up
+        ? Icons.add_circle_outline
+        : Icons.remove_circle_outline;
+    colorTween =
+        ColorTween(begin: lastColor, end: color).animate(colorController);
+    if (widget.type == VoteButtonType.favorite) {
+      icon = null;
+      favoriteMorphCurveAnimation = CurvedAnimation(
+          parent: favoriteController,
+          curve: Curves.easeOutCubic,
+          reverseCurve: Curves.easeInCubic);
+      favoriteMorph = Tween(
+              begin: widget.animationService.initialVote == Vote.favorite
+                  ? 1.0
+                  : 0.0,
+              end: widget.animationService.initialVote != Vote.favorite
+                  ? 1.0
+                  : 0.0)
+          .animate(favoriteMorphCurveAnimation);
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedBuilder(
+      animation: colorTween,
+      builder: (_, __) => RotationTransition(
+        turns: rotationController,
+        child: _buildBaseVoteButton(color: colorTween.value, icon: icon),
+      ),
+    );
+  }
+
+  IconButton _buildBaseVoteButton({Color color, IconData icon}) => IconButton(
```

Comment:
Use a function body and make use of some variables. This ternary operator stuff and inline function stuff is not readable well.